R/sensitivityCalculation.R

```diff
@@ -82,48 +82,131 @@ sensitivityCalculation <- function(simulation,
+    }
+
+    initialValues[[parameterPath]] <- param$value
+  }
+
+  constantParamPaths <- unlist(constantParamPaths)
+  formulaParamPaths <- unlist(formulaParamPaths)
+
+  simulationBatches <- list()
+
+  # Create one batch for all constant parameters
+  if (length(constantParamPaths) > 0) {
+    constantBatch <- createSimulationBatch(
+      simulation = simulation,
+      parametersOrPaths = constantParamPaths
     )
-  )
 
-  # name list with name of each parameter path
-  names(simulationResultsBatch) <- parameterPaths
+    # Add run values. While varying one parameter, the values of remaining
+    # constant parameters remain at their initial values.
+    for (constantParamPath in constantParamPaths) {
+      for (scaleFactorIdx in seq_along(variationRange)) {
+        # Change the value of the varied parameter
+        runValues <- initialValues[constantParamPaths]
+        runValues[[constantParamPath]] <- variationRange[[scaleFactorIdx]] * runValues[[constantParamPath]]
 
-  # extract and save dataframes ------------------------
+        # Add run values and store the ID in the `batchResultsIdMap`
+        batchResultsIdMap[[constantParamPath]][[scaleFactorIdx]] <- constantBatch$addRunValues(parameterValues = runValues)
+      }
+    }
 
-  # extract dataframe for PK parameters
+    simulationBatches <- c(simulationBatches, constantBatch)
+  }
+
+  # Add batches for formula parameters
+  for (formulaParamPath in formulaParamPaths) {
+    formulaBatch <- createSimulationBatch(
+      simulation = simulation,
+      parametersOrPaths = formulaParamPath
+    )
+
+    # Add run values.
+    for (scaleFactorIdx in seq_along(variationRange)) {
+      batchResultsIdMap[[formulaParamPath]][[scaleFactorIdx]] <- formulaBatch$addRunValues(parameterValues = variationRange[[scaleFactorIdx]] * initialValues[[formulaParamPath]])
+    }
+
+    simulationBatches <- c(simulationBatches, formulaBatch)
+  }
+
+  # Simulate all batches in parallel
+  simulationBatchesResults <- runSimulationBatches(simulationBatches = simulationBatches)
+
+  # `runSimulationBatches()` returns a list with one entry per simulation batch.
+  #
+  # Unlist so all results are in one list. The names of the results are the IDs
+  # of the runs. Using `batchResultsIdMap`, results for a certain
+  # parameter/scale factor combination can be filtered out.
+  simulationBatchesResults <- unlist(simulationBatchesResults)
+
+  # Create a nested list of simulation batch results.
+  # Outer list indexes each parameter path, while inner list corresponds to each
+  # value of parameter factor.
+  simulationResultsBatch <- batchResultsIdMap
+
+  for (parameterPath in seq_along(simulationResultsBatch)) {
+    for (parameterFactor in seq_along(simulationResultsBatch[[parameterPath]])) {
+      simulationResultsBatch[[parameterPath]][[parameterFactor]] <- purrr::pluck(simulationBatchesResults, batchResultsIdMap[[parameterPath]][[parameterFactor]])
+    }
+  }
```

Comment:
The new `simulationResultsBatch` is the same as the old one, and so everything from here on works the same way as before.

Reply:
Nice, that was easy!